tests/common.rs

```diff
@@ -1,482 +1,160 @@
-    );
-
-    c.close().await
-}
 
-async fn finds_sub_elements(mut c: Client) -> Result<(), error::CmdError> {
-    // Go to the Wikipedia front page
-    c.goto("https://en.wikipedia.org/").await?;
-    // Get the main sidebar panel
-    let mut panel = c.find(Locator::Css("div#mw-panel")).await?;
-    // Get all the ul elements in the sidebar
-    let mut portals = panel.find_all(Locator::Css("div.portal")).await?;
-
-    let portal_titles = &[
-        // Because GetElementText (used by Element::text()) returns the text
-        // *as rendered*, hidden elements return an empty String.
-        "",
-        "Interaction",
-        "Tools",
-        "In other projects",
-        "Print/export",
-        "Languages",
-    ];
-    // Unless something fundamentally changes, this should work
-    assert_eq!(portals.len(), portal_titles.len());
-
-    for (i, portal) in portals.iter_mut().enumerate() {
-        // Each "portal" has an h3 element.
-        let mut portal_title = portal.find(Locator::Css("h3")).await?;
-        let portal_title = portal_title.text().await?;
-        assert_eq!(portal_title, portal_titles[i]);
-        // And also an <ul>.
-        let list_entries = portal.find_all(Locator::Css("li")).await?;
-        assert!(!list_entries.is_empty());
-    }
+#[macro_export]
+macro_rules! tester {
+    // Ident should identify an async fn that takes just a Client.
+    ($f:ident, $endpoint:expr) => {{
+        use std::sync::{Arc, Mutex};
+        use std::thread;
+
+        let c = common::select_client_type($endpoint);
+
+        // we'll need the session_id from the thread
+        // NOTE: even if it panics, so can't just return it
+        let session_id = Arc::new(Mutex::new(None));
+
+        // run test in its own thread to catch panics
+        let sid = session_id.clone();
+        let res = thread::spawn(move || {
+            let mut rt = tokio::runtime::Builder::new().enable_all().basic_scheduler().build().unwrap();
+            let mut c = rt.block_on(c).expect("failed to construct test client");
+            *sid.lock().unwrap() = rt.block_on(c.session_id()).unwrap();
+            let x = rt.block_on($f(c));
+            drop(rt);
+            x
+        })
+        .join();
+        let success = common::handle_test_error(res);
+        assert!(success);
+    }};
 
-    c.close().await
 }
 
-async fn persist_inner(mut c: Client) -> Result<(), error::CmdError> {
-    c.goto("https://en.wikipedia.org/").await?;
-    c.persist().await?;
-
-    c.close().await
+/// Sets up the server and returns the port it bound to.
+pub fn setup_server() -> u16 {
+    let (tx, rx) = std::sync::mpsc::channel();
+
+    std::thread::spawn(move || {
+        let mut rt = tokio::runtime::Builder::new().enable_all().basic_scheduler().build().unwrap();
+        let _ = rt.block_on(async {
+            let (socket_addr, server) = start_server();
+            tx.send(socket_addr.port()).expect("To be able to send port");
+            server
```

Comment:
Shouldn't this be `server.await`?

Reply:
well... uuuh... that would explain a lot...